File: AgenticADB/agentic_adb/parser/adb_parser.py
import xml.etree.ElementTree as ET
import re
from agentic_adb.models import UIElement
from agentic_adb.exceptions import ParseError
from .base_parser import BaseParser

class ADBParser(BaseParser[UIElement]):
    """Parses an Android UI XML string and returns a filtered list of UI elements."""

    def parse(self, raw: str) -> list[UIElement]:
        try:
            root = ET.fromstring(raw)
        except ET.ParseError as e:
            raise ParseError(f"Failed to parse Android XML: {e}")

        elements = []
        filtered_index = 1

        for node in root.iter("node"):
            text = node.attrib.get("text", "")
            desc = node.attrib.get("content-desc", "")
            resource_id = node.attrib.get("resource-id", "")
            clickable_str = node.attrib.get("clickable", "false")
            clickable = clickable_str.lower() == "true"

            if not text and not desc and not resource_id and not clickable:
                continue

            class_name_full = node.attrib.get("class", "")
            class_name = class_name_full.split(".")[-1] if class_name_full else ""

            id_str = resource_id
            if "/" in id_str:
                id_str = id_str.split("/", 1)[-1]

            bounds_str = node.attrib.get("bounds", "")
            center_x, center_y = 0, 0
            if bounds_str:
                match = re.match(r"\[(\d+),(\d+)\]\[(\d+),(\d+)\]", bounds_str)
                if match:
                    x1, y1, x2, y2 = map(int, match.groups())
                    center_x = (x1 + x2) // 2
                    center_y = (y1 + y2) // 2

            element = UIElement(
                index=filtered_index,
                class_name=class_name,
                text=text,
                id=id_str,
                desc=desc,
                clickable=clickable,
                center_x=center_x,
                center_y=center_y,
            )
            elements.append(element)
            filtered_index += 1

        return elements

File: src/parser/adb-parser.ts
import { XMLParser } from "fast-xml-parser";
import { ParseError } from "../client/base-client.js";
import { createUIElement, type UIElement } from "../models/ui-element.js";

export interface BaseParser {
	parse(data: string): UIElement[];
}

export class AdbParser implements BaseParser {
	parse(xmlData: string): UIElement[] {
		if (!xmlData || xmlData.trim() === "") {
			return [];
		}

		let cleanXml = xmlData;
		const xmlStart = xmlData.indexOf("<?xml");
		if (xmlStart > 0) {
			cleanXml = xmlData.substring(xmlStart);
		} else if (xmlStart === -1 && xmlData.includes("<hierarchy")) {
			const hierarchyStart = xmlData.indexOf("<hierarchy");
			cleanXml = xmlData.substring(hierarchyStart);
		}

		try {
			const parser = new XMLParser({
				ignoreAttributes: false,
				attributeNamePrefix: "@_",
				isArray: (name) => name === "node",
			});
			const parsed = parser.parse(cleanXml);

			if (!parsed || !parsed.hierarchy) {
				throw new ParseError("Invalid XML format");
			}

			const elements: UIElement[] = [];
			let index = 0;

			const traverse = (node: any) => {
				if (!node) return;

				if (node["@_class"]) {
					const rawClass = node["@_class"] as string;
					const classParts = rawClass.split(".");
					const className = classParts[classParts.length - 1];

					const rawId = node["@_resource-id"] || "";
					let id = rawId;
					if (rawId && rawId.includes("/")) {
						id = rawId.split("/")[1];
					}

					const text = node["@_text"] || "";
					const desc = node["@_content-desc"] || "";
					const clickable = node["@_clickable"] === "true";
					const bounds = node["@_bounds"] || "";

					let cx = 0;
					let cy = 0;
					let validBounds = false;
					const match = bounds.match(/\[(\d+),(\d+)\]\[(\d+),(\d+)\]/);
					if (match) {
						const x1 = parseInt(match[1], 10);
						const y1 = parseInt(match[2], 10);
						const x2 = parseInt(match[3], 10);
						const y2 = parseInt(match[4], 10);
						cx = Math.floor((x1 + x2) / 2);
						cy = Math.floor((y1 + y2) / 2);
						validBounds = true;
					}

					if (text || desc || id || clickable) {
						elements.push(
							createUIElement({
								index: index++,
								class: className,
								text,
								id,
								desc,
								clickable,
								center_x: validBounds ? cx : 0,
								center_y: validBounds ? cy : 0,
							}),
						);
					}
				}

				if (node.node) {
					const children = Array.isArray(node.node) ? node.node : [node.node];
					for (const child of children) {
						traverse(child);
					}
				}
			};

			traverse(parsed.hierarchy);
			return elements;
		} catch (e: any) {
			if (e instanceof ParseError) throw e;
			throw new ParseError(`Failed to parse Android UI XML: ${e.message}`);
		}
	}
}
